### PR TITLE
[GH-15579] migrate lifecycle methods from SearchBar component

### PR DESCRIPTION
**Pull request by ArcaneDiver**
#### Summary
This PR migrate the lifecycle method `componentWillReceiveProps` to `getDerivedState`

#### Ticket Link
Fixes https://github.com/mattermost/mattermost-server/issues/15579

### CODE REVIEW

File: components/search_bar/search_bar.jsx
Comment (hanzei) [inline components/search_bar/search_bar.jsx:90]: Heads up that CI reports two linter errors: /root/mattermost/mattermost-webapp/components/search_bar/search_bar.jsx
  89:95  error  Missing trailing comma  comma-dangle
  90:10  error  Missing semicolon       semi

Comment (ArcaneDiver) [reply components/search_bar/search_bar.jsx]: Ok i just did a commit that fixes the problems

File: components/search_bar/search_bar.jsx
Comment (hmhealey) [inline components/search_bar/search_bar.jsx:84]: I think this needs to be `getDerivedStateFromProps`

File: components/search_bar/search_bar.jsx
Comment (hmhealey) [inline components/search_bar/search_bar.jsx:88]: You actually don't need to return unchanged state from `getDerivedStateFromProps`, so this can be removed. If nothing is changing in the state, you just return null.

**Review by hmhealey — CHANGES_REQUESTED** [review]
Thanks for the PR, @ArcaneDiver. I have a few suggestions about how to use `getDerivedStateFromProps` a bit better.

File: components/search_bar/search_bar.jsx
Comment (ArcaneDiver) [reply components/search_bar/search_bar.jsx:63]: Ok i'll do it

File: components/search_bar/search_bar.jsx
Comment (ArcaneDiver) [reply components/search_bar/search_bar.jsx:84]: Ah, that was an huge mistake 😅.

File: components/search_bar/search_bar.jsx
Comment (ArcaneDiver) [reply components/search_bar/search_bar.jsx:89]: Ok, i'll save last searchTerms in a static field.

File: components/search_bar/search_bar.jsx
Comment (ArcaneDiver) [reply components/search_bar/search_bar.jsx]: I just pushed a modify that track last changed prop in order to recompute `visibleSearchHintOptions` only when necessary.

**Comment by ArcaneDiver** [comment]
Hey i just re added 
```js
this.state = {
            focused: false,
            keepInputFocused: false,
            highlightedSearchHintIndex: -1,
--->     visibleSearchHintOptions: SearchBar.determineVisibleSearchHintOptions(props.searchTerms),
};
```
because without setting `visibleSearchHintOptions` in the constructor tests did not pass.

**Comment by ArcaneDiver** [comment]
I can not understand why build-docker fails...

**Review by nevyangelova — APPROVED** [review]

**Comment by hmhealey** [comment]
Looks like the `build-docker` failure was just a random CircleCI issue. I restarted the build for that

File: components/search_bar/search_bar.jsx
Comment (hmhealey) [inline components/search_bar/search_bar.jsx:90]: Thanks for making those changes.

Could you also put the `lastSearchTerms` into the state? That's consistent with how we do it elsewhere. Having it in the state instead of as a static field is a bit more idiomatic, and it prevents multiple instances of the component with interfering with each other.

**Review by hmhealey — CHANGES_REQUESTED** [review]

File: components/search_bar/search_bar.jsx
Comment (ArcaneDiver) [reply components/search_bar/search_bar.jsx]: Here it is the new commit!

**Review by hmhealey — APPROVED** [review]
Thanks! Everything looks good to me now

**Comment by ArcaneDiver** [comment]
I just pull from upstream. I have to do something else before merging?

**Comment by jasonblais** [comment]
@ArcaneDiver this is just pending final QA review from @jgilliam17 and then we're good to merge :+1:

**Review by jgilliam17 — APPROVED** [review]
Thank you @ArcaneDiver - apologize about delay in getting to this PR
Tested, looks good to merge.
- Verified/tested around search bar, focus, hint options etc. - as expected.
@hanzei Can you please help merge? Thanks.

**Pull request merged**
